pulp_container/app/global_access_conditions.py

```diff
@@ -0,0 +1,221 @@
+    * "file.view_fileremote" - Permission to view the ``FileRemote``.
+    * "file.delete_fileremote" - Permission to delete the ``FileRemote``.
+
+    Args:
+        request (rest_framework.request.Request): The request being made.
+        view (subclass rest_framework.viewsets.GenericViewSet): The view being checked for
+            authorization.
+        action (str): The action being performed, e.g. "destroy".
+        permission (str): The name of the Permission to be checked. In the form
+            `app_label.codename`, e.g. "core.delete_task".
+
+    Returns:
+        True if the user has the Permission named by the ``permission`` argument on the ``remote``
+            parameter at the object-level. False otherwise.
+
+    """
+    serializer = RepositorySyncURLSerializer(data=request.data, context={"request": request})
+    serializer.is_valid(raise_exception=True)
+    remote = serializer.validated_data.get("remote")
+    return request.user.has_perm(permission, remote)
+
+
+def has_remote_param_model_or_obj_perms(request, view, action, permission):
+    """
+    Checks if the current user has either model-level or object-level permissions on the ``remote``.
+
+    The object in this case is the one specified by the ``remote`` parameter. For example when
+    syncing the ``remote`` parameter is passed in as an argument.
+
+    This is usable as a conditional check in an AccessPolicy. Here is an example checking for
+    "file.view_fileremote" permission at either the model-level or object-level.
+
+    ::
+
+        {
+            ...
+            "condition": "has_remote_param_model_or_obj_perms:file.view_fileremote",
+        }
+
+    Since it is checking a ``remote`` object the permission argument should be one of the following:
+
+    * "file.change_fileremote" - Permission to change the ``FileRemote``.
+    * "file.view_fileremote" - Permission to view the ``FileRemote``.
+    * "file.delete_fileremote" - Permission to delete the ``FileRemote``.
+
+    Args:
+        request (rest_framework.request.Request): The request being made.
+        view (subclass rest_framework.viewsets.GenericViewSet): The view being checked for
+            authorization.
+        action (str): The action being performed, e.g. "destroy".
+        permission (str): The name of the Permission to be checked. In the form
+            `app_label.codename`, e.g. "core.delete_task".
+
+    Returns:
+        True if the user has the Permission named by the ``permission`` at the model-level or on the
+            argument on the ``remote`` parameter at the object-level. False otherwise.
+
+    """
+    if has_model_perms(request, view, action, permission):
+        return True
+    if has_remote_param_obj_perms(request, view, action, permission):
+        return True
+    return False
+
+
+# Custom checks for namespaces within distributions
+
+
+def has_add_namespace_perms(request, view, action):
```

Comment:
at the next step we need to figure out how to enable this situation
```
$ http POST  :24817/pulp/api/v3/distributions/container/container/ base_path=library/production name=distribute-rhel7 --auth ina:goodpassword
HTTP/1.1 202 Accepted
Access-Control-Expose-Headers: Correlation-ID
Allow: GET, POST, HEAD, OPTIONS
Connection: close
Content-Length: 67
Content-Type: application/json
Correlation-ID: 878cf1b9197345e3a1240f01f3c671f7
Date: Tue, 22 Dec 2020 12:12:41 GMT
Server: gunicorn/20.0.4
Vary: Accept, Cookie
X-Frame-Options: SAMEORIGIN

{
    "task": "/pulp/api/v3/tasks/d0357850-40e7-4d56-8ad0-dbc050374c50/"
}

(pulp) [vagrant@pulp3-source-fedora32 _scripts]$ http POST  :24817/pulp/api/v3/distributions/container/container/ base_path=library/dev name=dev-rhel7 --auth lena:badpassword
HTTP/1.1 403 Forbidden
Access-Control-Expose-Headers: Correlation-ID
Allow: GET, POST, HEAD, OPTIONS
Connection: close
Content-Length: 63
Content-Type: application/json
Correlation-ID: 420b9be882a847448c57bfc5431557e4
Date: Tue, 22 Dec 2020 12:13:40 GMT
Server: gunicorn/20.0.4
Vary: Accept, Cookie
X-Frame-Options: SAMEORIGIN

{
    "detail": "You do not have permission to perform this action."
}

```
Both users have add namespace and add distribution perms assigned, however due to queryset scoping thay can perform operations only on the objects they have created.

Reply:
With the current changes, your scenario is handled correctly. As I mentioned in our meeting, we should consider both the model-level permissions along with object-level permission at the same time. In particular, this had to be modified:
```diff
- return request.user.has_perm("container.view_containernamespace", namespace)
+ return request.user.has_perm(
+     "container.view_containernamespace"
+ ) or request.user.has_perm("container.view_containernamespace", namespace)
```